Algorithm/QCAlgorithm.Trading.cs

```diff
@@ -604,36 +607,32 @@ public int CalculateOrderQuantity(Symbol symbol, decimal target)
             // determine the unit price in terms of the account currency
             var unitPrice = new MarketOrder(symbol, 1, UtcTime).GetValue(security);
 
-            // define lower and upper thresholds for the iteration
-            var lowerThreshold = targetOrderValue - unitPrice / 2;
-            var upperThreshold = targetOrderValue + unitPrice / 2;
+            // calculate the total margin available
+            var marginRemaining = Portfolio.GetMarginRemaining(symbol, direction);
+            if (marginRemaining <= 0) return 0;
 
-            // continue iterating while  we're still not within the specified thresholds
-            var iterations = 0;
+            // continue iterating while we do not have enough margin for the order
+            decimal marginRequired;
             var orderQuantity = 0;
-            decimal orderValue = 0;
-            while ((orderValue < lowerThreshold || orderValue > upperThreshold) && iterations < 10)
-            {
-                // find delta from where we are to where we want to be
-                var delta = targetOrderValue - orderValue;
-                // use delta value to compute a change in quantity required
-                var deltaQuantity = (int)(delta / unitPrice);
 
-                orderQuantity += deltaQuantity;
+            do
+            {
+                if (orderQuantity == 0)
+                {
+                    // compute the initial order quantity
+                    orderQuantity = (int)(targetOrderValue / unitPrice);
+                }
+                else
+                {
+                    // decrease the order quantity
+                    orderQuantity--;
+                }
 
-                // recompute order fees
+                // calculate the margin required for the order
                 var order = new MarketOrder(security.Symbol, orderQuantity, UtcTime);
-                var fee = security.FeeModel.GetOrderFee(security, order);
-
-                orderValue = Math.Abs(order.GetValue(security)) + fee;
+                marginRequired = security.MarginModel.GetInitialMarginRequiredForOrder(security, order);
 
-                // we need to add the fee in as well, even though it's not value, it's still a cost for the transaction
-                // and we need to account for it to be sure we can make the trade produced by this method, imagine
-                // set holdings 100% with 1x leverage, but a high fee structure, it quickly becomes necessary to include
-                // otherwise the result of this function will be inactionable.
-
-                iterations++;
-            }
+            } while (marginRequired > marginRemaining || marginRequired > targetOrderValue);
```

Comment:
You are right, the number of iterations for the MSFT tests with 10K constant fee is 401 :(
One possible solution could be decrementing by the orderFee/price ratio instead of by one.
With large constant fees this should give the correct solution at the 2nd iteration, still not sure about the non-linear case you mentioned (more tests needed).

One question: why is the initial value of `orderQuantity` an "estimate" ? Isn't it always equal to the upper-bound, which is also the correct value in the absence of fees ?